base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs

```diff
@@ -2546,13 +2586,72 @@ impl BlockchainBackend for LMDBDatabase {
         Ok(chain_header)
     }
 
-    /// Returns the metadata of the chain.
+    // Returns the metadata of the chain.
     fn fetch_chain_metadata(&self) -> Result<ChainMetadata, ChainStorageError> {
         let txn = self.read_transaction()?;
         let metadata = fetch_metadata(&txn, &self.metadata_db)?;
         Ok(metadata)
     }
 
+    // Returns the payref rebuild status.
+    fn fetch_payref_rebuild_status(&self) -> Result<PayrefRebuildStatus, ChainStorageError> {
+        let txn = self.read_transaction()?;
+
+        let val: Option<MetadataValue> = lmdb_get(&txn, &self.metadata_db, &MetadataKey::PayrefRebuildStatus.as_u32())?;
+        match val {
+            Some(MetadataValue::PayrefRebuildStatus(status)) => Ok(status),
+            _ => Ok(PayrefRebuildStatus::default()),
+        }
+    }
+
+    // Builds the payref indexes for a given block height, with stats.
+    fn build_payref_indexes_for_height(
+        &self,
+        height: u64,
+        metadata_at_start: ChainMetadata,
+        initialize_stats: Option<u64>,
+        finalize: bool,
+    ) -> Result<PayrefRebuildStatus, ChainStorageError> {
+        let write_txn = self.write_transaction()?;
+        if let Some(height) = initialize_stats {
+            self.set_stats_total_height(height);
+        }
+
+        // Get all outputs for this block
+        let binding = self.fetch_chain_header_by_height(height)?;
+        let header = binding.header();
+        let block_hash = header.hash();
+        let outputs = self.fetch_outputs_in_block(&block_hash)?;
+
+        // Update the payref index for each output
+        for output in &outputs {
+            let output_hash = output.hash();
+            self.update_payref(&write_txn, &block_hash, &output_hash)?;
+        }
```

Comment:
```suggestion
        let outputs : Vec<(Vec<u8>, TransactionOutputRowData)> =
        lmdb_fetch_matching_after(&write_txn, &db.utxos_db, block_hash.as_slice())?;

        // Update the payref index for each output
        for output in &outputs {
            self.update_payref(&write_txn, &block_hash, &output.hash)?;
        }
```
Then you don't have to hash them, that data is already in the db